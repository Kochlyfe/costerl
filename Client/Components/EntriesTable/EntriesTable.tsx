import React from 'react';
import { StyleSheet, View, Text, TouchableOpacity, Alert } from 'react-native';
import { Table, TableWrapper, Row, Cell } from 'react-native-table-component';
import { Entry } from '../../interfaces';

interface DeleteOne {
  (id: number): void;
}

interface Props {
  deleteOne: DeleteOne;
  userEntries: Entry[];
}

const EntriesTable = ({ userEntries, deleteOne }: Props): JSX.Element => {
  const tableHead = ['id', 'item', 'category', 'amount', 'delete'];

  const element = (id: number): JSX.Element => (
    <TouchableOpacity
      onPress={() => {
        Alert.alert(`The item has been deleted`);
        deleteOne(id);
      }}
    >
      <View style={styles.btn}>
        <Text style={styles.btnText}>ðŸ—‘</Text>
      </View>
    </TouchableOpacity>
  );

  const tableRender = (): JSX.Element => {
    return (
      <Table borderStyle={{ borderColor: 'transparent' }}>
        <Row data={tableHead} style={styles.head} textStyle={styles.text} />
        {userEntries.map((entry: Entry, index: number) => {
          return (
            <TableWrapper key={index} style={styles.row}>
              {['id', 'item', 'category', 'amount', 'delete'].map((cellData, cellIndex: number) => (
                <Cell
                  key={cellIndex}
                  data={cellIndex === 4 ? element(cellIndex) : entry[cellData]}
                  textStyle={styles.text}
                />
              ))}
            </TableWrapper>
          );
        })}
      </Table>
    );
  };

  return <>{Array.isArray(userEntries) ? tableRender() : <></>}</>;
};

export default EntriesTable;

const styles = StyleSheet.create({
  container: { flex: 1, padding: 16, paddingTop: 30, backgroundColor: '#fff' },
  head: { height: 40, backgroundColor: '#2aa198', marginBottom: 5 },
  text: { margin: 6, color: 'white' },
  row: { flexDirection: 'row', backgroundColor: '#268bd2', marginBottom: 10 },
  btn: { width: 58, height: 18, backgroundColor: '#268bd2' },
  btnText: { textAlign: 'center', color: '#fff' },
});
