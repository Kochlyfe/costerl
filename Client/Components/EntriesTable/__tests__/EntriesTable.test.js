import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import EntriesTable from '../EntriesTable.tsx';

const mockEntries = [
  {
    amount: '1',
    category: 'Food',
    createdAt: '2020-11-20T10:44:18.662Z',
    currency: 'USD',
    date: '2020-11-20T10:44:18.662Z',
    description: 'Really good pizza',
    flag: true,
    id: 1,
    item: 'Pizza',
    payment: 5,
    updatedAt: '2020-11-20T10:44:18.662Z',
    userId: 1,
  },
  {
    amount: '2',
    category: 'Food',
    createdAt: '2020-11-20T10:44:18.662Z',
    currency: 'USD',
    date: '2020-11-20T10:44:18.662Z',
    description: 'Really good risotto',
    flag: true,
    id: 1,
    item: 'Risotto',
    payment: 10,
    updatedAt: '2020-11-20T10:44:18.662Z',
    userId: 1,
  },
];

describe('EntriesTable', () => {
  test('snapshot', () => {
    const { toJSON } = render(<EntriesTable userEntries={mockEntries} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('displays table header when input is empty', () => {
    const { getByText } = render(<EntriesTable userEntries={[]} />);
    expect(getByText('id')).toBeTruthy();
    expect(getByText('item')).toBeTruthy();
    expect(getByText('category')).toBeTruthy();
    expect(getByText('amount')).toBeTruthy();
    expect(getByText('delete')).toBeTruthy();
  });

  test('displays table header when input not empty', () => {
    const { getByText } = render(<EntriesTable userEntries={mockEntries} />);
    expect(getByText('id')).toBeTruthy();
    expect(getByText('item')).toBeTruthy();
    expect(getByText('category')).toBeTruthy();
    expect(getByText('amount')).toBeTruthy();
    expect(getByText('delete')).toBeTruthy();
  });

  test('displays correct entries', () => {
    const { getByText, getAllByText } = render(<EntriesTable userEntries={mockEntries} />);
    expect(getAllByText('Food').length).toBe(2);
    expect(getByText('Pizza')).toBeTruthy();
    expect(getByText('Risotto')).toBeTruthy();
  });

  test('calls delete button', () => {
    const mockOnClick = jest.fn();
    const { getAllByText } = render(<EntriesTable userEntries={mockEntries} deleteOne={mockOnClick} />);
    fireEvent(getAllByText('ðŸ—‘')[0], 'click', { preventDefault: jest.fn() });
  });
});
