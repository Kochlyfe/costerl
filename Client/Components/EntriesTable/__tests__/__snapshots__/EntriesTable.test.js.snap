// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntriesTable snapshot 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "borderBottomWidth": 0,
        "borderColor": "transparent",
        "borderLeftWidth": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        0,
        Object {
          "flexDirection": "row",
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#2aa198",
          "height": 40,
          "marginBottom": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        id
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        item
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        category
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        amount
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        delete
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#268bd2",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        1
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Pizza
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Food
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#268bd2",
              "height": 18,
              "width": 58,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "textAlign": "center",
              }
            }
          >
            ðŸ—‘
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#268bd2",
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        1
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Risotto
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Food
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "margin": 6,
            },
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "transparent",
            "borderRightWidth": 0,
            "borderTopWidth": 0,
          },
          Object {
            "justifyContent": "center",
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#268bd2",
              "height": 18,
              "width": 58,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "textAlign": "center",
              }
            }
          >
            ðŸ—‘
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
